File: javascript/35.search-insert-position.js
/*
 * @lc app=leetcode id=35 lang=javascript
 *
 * [35] Search Insert Position
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function (nums, target) {
  let left = 0,
    right = nums.length - 1;
  if (target <= nums[left]) return left;
  else if (target === nums[right]) return right;
  else if (target > nums[right]) return right + 1;
  while (left < right) {
    const middle = ~~((left + right) / 2);
    if (nums[middle] >= target) right = middle;
    else if (nums[middle] < target) {
      left = middle + 1;
    }
  }
  return left;
};
// @lc code=end

File: rust/35.search-insert-position.rs
/*
 * @lc app=leetcode id=35 lang=rust
 *
 * [35] Search Insert Position
 */

// @lc code=start
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        return nums.binary_search(&target).unwrap_or_else(|x| x) as i32;

        // let (mut left, mut right) = (0, nums.len());
        // while left < right {
        //     let mid = (left + right) / 2;
        //     if nums[mid] < target {
        //         left = mid + 1;
        //     } else {
        //         right = mid;
        //     }
        // }
        // return right as i32;
    }
}
// @lc code=end
